File: Frontend/src/app/core/services/student.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

import { environment } from '../../../environments/env';
import { StudentSummary } from '../models/summaryResponse.dto';
import { StudentDetailedResponse } from '../models/student-verification.dto/student-verification.dto';

export interface CompleteStudentProfileRequest {
  instituteId: string;
  courseId: string;
  categoryId: string;
  addressId: string;
  location: string | null;
  pincode: number;
  aadharNumber: string;
  gender: string;
  dateOfBirth: string;
  fatherName: string;
  motherName: string;
  annualFamilyIncome: number;
}

export interface UpdateMyStudentProfileRequest extends Omit<CompleteStudentProfileRequest, 'aadharNumber'> {
  aadharNumber: string | null;
}

export interface CourseOption {
  id: string;
  name: string;
}

@Injectable({
  providedIn: 'root'
})
export class StudentService {

  private readonly API_URL = `${environment.apiUrl}/student`;

  constructor(private readonly http: HttpClient) {}

  getStudents(): Observable<StudentSummary[]> {
    return this.http.get<StudentSummary[]>(`${this.API_URL}/all`);
  }

  getStudentById(studentId: string): Observable<StudentDetailedResponse> {
    return this.http.get<StudentDetailedResponse>(`${this.API_URL}/${studentId}`);
  }

  getMyProfile(): Observable<StudentDetailedResponse> {
    return this.http.get<StudentDetailedResponse>(`${this.API_URL}/me`);
  }

  getCoursesByInstitute(instituteId: string): Observable<CourseOption[]> {
    return this.http.get<CourseOption[]>(`${environment.apiUrl}/courses/institute/${instituteId}`);
  }

  completeProfile(request: CompleteStudentProfileRequest): Observable<void> {
    return this.http.post<void>(`${this.API_URL}/complete-profile`, request);
  }

  updateMyProfile(request: UpdateMyStudentProfileRequest): Observable<void> {
    return this.http.put<void>(`${this.API_URL}/me/profile`, request);
  }
}

File: Backend/src/main/java/com/vidyasahay/vidyasahay/service/StudentService.java
package com.vidyasahay.vidyasahay.service;

import java.util.List;
import java.util.UUID;

import com.vidyasahay.vidyasahay.dto.request.CompleteStudentProfileRequest;
import com.vidyasahay.vidyasahay.dto.request.UpdateMyStudentProfileRequest;
import com.vidyasahay.vidyasahay.dto.response.StudentDetailedResponse;
import com.vidyasahay.vidyasahay.dto.response.StudentSummaryResponse;

public interface StudentService {

    List<StudentSummaryResponse> getAllStudents();

    List<StudentSummaryResponse> getStudentsForInstitute(UUID instituteUserId);

    StudentDetailedResponse getStudentById(UUID studentId);

    StudentDetailedResponse getStudentByUserId(UUID userId);

    void completeProfile(UUID userId, CompleteStudentProfileRequest request);

    StudentDetailedResponse updateMyProfile(UUID userId, UpdateMyStudentProfileRequest request);
}
